src/interpreter.c

```diff
@@ -152,16 +152,22 @@ void jl_reset_instantiate_inner_types(jl_datatype_t *t);
 
 void jl_set_datatype_super(jl_datatype_t *tt, jl_value_t *super)
 {
-    if (!jl_is_datatype(super) || !jl_is_abstracttype(super) ||
-        tt->name == ((jl_datatype_t*)super)->name ||
-        jl_subtype(super, (jl_value_t*)jl_vararg_type) ||
-        jl_is_tuple_type(super) ||
-        jl_is_namedtuple_type(super) ||
-        jl_subtype(super, (jl_value_t*)jl_type_type) ||
-        jl_subtype(super, (jl_value_t*)jl_builtin_type)) {
-        jl_errorf("invalid subtyping in definition of %s",
-                  jl_symbol_name(tt->name->name));
-    }
+    if (!jl_is_datatype(super)|| !jl_is_abstracttype(super)) jl_type_error("Definition of supertype",
+    	(jl_value_t*)jl_datatype_type, super);  
+
+    if (tt->name == ((jl_datatype_t*)super)->name) jl_errorf("Invalid supertyping in definition of %s:" 
+    	"A type cannot subtype itself.",jl_symbol_name(tt->name->name));  
+    
+    if (jl_subtype(super, (jl_value_t*)jl_type_type)) jl_errorf("Invalid supertyping in definition of %s:"
+        "Cannot add subtypes to Core.Type.",jl_symbol_name(tt->name->name));
+    
+    if (jl_subtype(super, (jl_value_t*)jl_builtin_type)) jl_errorf("Invalid supertyping in definition of %s:"
+    	"Cannot add subtypes to Core.Builtin.",jl_symbol_name(tt->name->name));
+   
+    if (jl_subtype(super, (jl_value_t*)jl_vararg_type)) jl_errorf("Invalid supertyping in definition of %s:"
+       	"Cannot add subtypes to Core.Vararg.",jl_symbol_name(tt->name->name));
+
+
```

Comment:
The code formatting needed fixing, but I think this can be significantly DRYer too.
```suggestion
    JL_TYPECHKS("declared supertype", datatype, super);
    const char *error = NULL;
    if (!jl_is_abstracttype(super))
        error = "Can only subtype abstract types";
    else if (tt->name == ((jl_datatype_t*)super)->name)
        error = "A type cannot subtype itself";
    else if (jl_subtype(super, (jl_value_t*)jl_type_type))
        error = "Cannot add subtypes to Type";
    else if (jl_subtype(super, (jl_value_t*)jl_builtin_type))
        error = "Cannot add subtypes to Core.Builtin";
    if (error)
         jl_errorf("Invalid supertyping in definition of %s: %s.", jl_symbol_name(tt->name->name), error);
```